nshlib/nsh_parse.c

```diff
@@ -1543,111 +1697,78 @@ static FAR char *nsh_argument(FAR struct nsh_vtbl_s *vtbl,
-        }
-      else
-        {
-          /* No, then any of the usual separators will terminate the
-           * argument.  In this case, pbegin points for the first character
-           * of the token following the previous separator.
-           */
-
-          term = g_token_separator;
-        }
-
-      /* Find the end of the string */
-
-#ifdef CONFIG_NSH_CMDPARMS
-      /* Some special care must be exercised to make sure that we do not
-       * break up any back-quote delimited substrings.  NOTE that the
-       * absence of a closing back-quote is not detected;  That case should
-       * be detected later.
+      /* Find the end of the string. Some special care must be exercised to
+       * make sure that we do not break up any quoted substrings.
        */
 
+      quoted = false;
 #ifdef CONFIG_NSH_QUOTE
-      quoted    = false;
-      backquote = false;
+      escaped = false;
 
       for (prev = NULL, pend = pbegin; *pend != '\0'; prev = pend, pend++)
+#else
+      for (pend = pbegin; *pend != '\0'; pend++)
+#endif
         {
-          /* Check if the current character is quoted */
+#ifdef CONFIG_NSH_QUOTE
+          /* Check if the current character is escaped */
 
-          if (prev != NULL && *prev == '\\' && !quoted)
+          if (prev != NULL && *prev == '\\' && !escaped)
             {
               /* Do no special checks on the quoted character */
 
-              quoted = true;
+              escaped = true;
               continue;
             }
 
-          quoted = false;
+          escaped = false;
 
-          /* Check if the current character is an (unquoted) back-quote */
+          /* Check if the current character is an (unescaped) back-slash */
 
-          if (*pend == '\\' && !quoted)
+          if (*pend == '\\' && !escaped)
             {
               /* Yes.. Do no special processing on the backspace character */
 
               continue;
             }
+#endif
 
-          /* Toggle the back-quote flag when one is encountered? */
+          /* Are we entering a quoted string ? */
 
-          if (*pend == '`')
+          if ((quoted = (nsh_strchr(g_quote_separator, *pend) != NULL)))
             {
-              backquote = !backquote;
-            }
+              /* Yes, find the terminator and continue from there */
 
-          /* Check for a delimiting character only if we are not in a
-           * back-quoted sub-string.
-           */
+              FAR char *qend = nsh_strchr(pend + 1, *pend);
+              if (!qend)
+                {
+                  /* No terminator found, get out */
 
-          else if (!backquote && nsh_strchr(term, *pend) != NULL)
-            {
-              /* We found a delimiter outside of any back-quoted substring.
-               * Now we can break out of the loop.
-               */
+                  char qterm[2];
 
-              break;
-            }
-        }
-#else
-      backquote = false;
+                  qterm[0] = ptr;
```

Comment:
Copy&paste error when fixing code style issue. Will fix this on monday (should be qterm[0] = *pend;)